File: frontend/src/types/request.ts
import { z } from "zod";

/*
 * The API contract, expressed once as Zod schemas. Everything the app knows
 * about a request's shape is inferred from here (`z.infer`), so the static
 * types and the runtime validation can never drift apart.
 *
 * The case lists four statuses; the backend adds `cancelled` as its own value
 * (documented in backend/README.md) so a user-triggered stop is not confused
 * with a real processing failure. The union therefore carries five values.
 */
export const statusSchema = z.enum([
  "pending",
  "processing",
  "completed",
  "error",
  "cancelled",
]);

export type Status = z.infer<typeof statusSchema>;

/** Statuses from which nothing else will happen — polling can stop here. */
export const TERMINAL_STATUSES: readonly Status[] = [
  "completed",
  "error",
  "cancelled",
];

export function isTerminal(status: Status): boolean {
  return TERMINAL_STATUSES.includes(status);
}

/** A request still eligible for the cancel action. */
export function isCancellable(status: Status): boolean {
  return status === "pending" || status === "processing";
}

/** Row returned by `GET /requests` — identity and progress, no log history. */
export const requestSummarySchema = z.object({
  id: z.string(),
  status: statusSchema,
  progress: z.number(),
  numbers: z.array(z.number()),
  result: z.number().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type RequestSummary = z.infer<typeof requestSummarySchema>;

/** Full object returned by `GET /requests/{id}` — the summary plus logs. */
export const requestDetailSchema = requestSummarySchema.extend({
  logs: z.array(z.string()),
});

export type RequestDetail = z.infer<typeof requestDetailSchema>;

/** Body of `POST /requests` — answers immediately, before processing runs. */
export const requestAcceptedSchema = z.object({
  id: z.string(),
  status: statusSchema,
});

export type RequestAccepted = z.infer<typeof requestAcceptedSchema>;

/** Confirmation returned by `POST /requests/{id}/cancel`. */
export const cancelResponseSchema = z.object({
  id: z.string(),
  status: statusSchema,
  message: z.string(),
});

export type CancelResponse = z.infer<typeof cancelResponseSchema>;

export const requestListSchema = z.array(requestSummarySchema);

File: backend/app/schemas/request.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field

MAX_NUMBERS = 1000


class Status(str, Enum):
    """Lifecycle of a request.

    ``cancelled`` is an addition of ours (the case allows "either error or a
    cancelled status of your own"), so a user-triggered stop is not confused
    with a real processing failure.
    """

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    ERROR = "error"
    CANCELLED = "cancelled"


#: Statuses from which nothing else can happen. Once a request reaches one of
#: these it is frozen: neither the processor nor a cancel call may change it.
FINAL_STATUSES = frozenset({Status.COMPLETED, Status.ERROR, Status.CANCELLED})


class RequestCreate(BaseModel):
    """Body of ``POST /requests`` — the form data, i.e. the list of numbers."""

    numbers: Annotated[list[float], Field(min_length=1, max_length=MAX_NUMBERS)]

    model_config = ConfigDict(json_schema_extra={"examples": [{"numbers": [10, 20, 5]}]})


class RequestSummary(BaseModel):
    """Row of ``GET /requests``: identity and progress, without the log history."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    status: Status
    progress: int
    numbers: list[float]
    result: float | None
    created_at: datetime
    updated_at: datetime


class RequestDetail(RequestSummary):
    """Full object returned by ``GET /requests/{id}``, logs included."""

    logs: list[str]


class RequestAccepted(BaseModel):
    """Body of ``POST /requests``: what the caller needs to start polling."""

    id: str
    status: Status


class CancelResponse(BaseModel):
    """Confirmation of the status change made by ``POST /requests/{id}/cancel``."""

    id: str
    status: Status
    message: str
